js/repl/Monaco.tsx

```diff
@@ -0,0 +1,228 @@
+import { css, type CSSInterpolation } from "@emotion/css";
+import * as monaco from "monaco-editor";
+import React, { useRef, useEffect } from "react";
+import { shikiToMonaco } from "@shikijs/monaco";
+import { createHighlighterCoreSync } from "shiki/core";
+import { createOnigurumaEngine } from "shiki/engine/oniguruma";
+import shikiTS from "@shikijs/langs/typescript";
+import shikiDarkPlus from "@shikijs/themes/dark-plus";
+import shikiLightPlus from "@shikijs/themes/light-plus";
+import shikiWasm from "@shikijs/engine-oniguruma/wasm-inlined";
+
+import { colors } from "./styles";
+import { preferDarkColorScheme } from "./Utils";
+
+type Props = {
+  className: string;
+  code: string;
+  placeholder: string;
+  onChange?: (value: string) => void;
+  fileSize: string | boolean;
+  lineWrapping: boolean;
+  errorMessage: string;
+  fastMode?: boolean;
+};
+
+const shikiPromise = (async function () {
+  const engine = await createOnigurumaEngine(shikiWasm);
+
+  shikiToMonaco(
+    createHighlighterCoreSync({
+      themes: [shikiDarkPlus, shikiLightPlus],
+      langs: [shikiTS],
+      engine: engine,
+    }),
+    monaco
+  );
+})();
+
+monaco.languages.typescript.typescriptDefaults.setCompilerOptions({
+  jsx: monaco.languages.typescript.JsxEmit.React,
+});
+
+export default function MonacoWithShiki(props: Props) {
+  const [asyncLoaded, setAsyncLoaded] = React.useState(false);
+  React.useEffect(() => {
+    shikiPromise.then(() => {
+      setAsyncLoaded(true);
+    });
+  }, []);
+  if (!asyncLoaded) {
+    return null;
+  }
+
+  return <Monaco {...props} />;
+}
+
+function Monaco({
+  className,
+  code,
+  placeholder,
+  onChange,
+  fileSize,
+  lineWrapping,
+  errorMessage,
+  fastMode,
+}: Props) {
+  const container = useRef<HTMLDivElement>(null);
+  // eslint-disable-next-line prefer-const
+  let [editor, setEditor] =
+    React.useState<monaco.editor.IStandaloneCodeEditor | null>(null);
+  const [rect, setRect] = React.useState<DOMRect | null>(null);
+
+  useEffect(() => {
+    setEditor(
+      (editor = monaco.editor.create(container.current, {
+        padding: {
+          top: 2,
+        },
+        // https://github.com/microsoft/monaco-editor/issues/4311
+        // automaticLayout: true,
+        model: fastMode
+          ? monaco.editor.createModel(
+              "",
+              "javascript",
+              monaco.Uri.file("output/output.tsx")
+            )
```

Comment:
This is a somewhat bad hack.

The syntax highlighter registered by shiki will cause performance degradation, which is obvious when using `timeTravel` on large files.
And monaco does not support using different syntax highlighters for the same language.
So we set the language of the output code to `js`, and notice that it will not warn even if we output ts, but there is no syntax highlighting for ts keywords.

What do you think about using this solution, or not using it (because it is not very slow, just a little slow), or continuing to use codemirror for output code?

Reply:
The current main is not significantly faster ([repl](https://babel.dev/repl#?browsers=defaults%2C%20not%20ie%2011%2C%20not%20ie_mob%2011&build=&builtIns=false&corejs=3.21&spec=false&loose=false&code_lz=IYZwngdgxgBAZgV2gFwJYHsI2ACgJQwDeAUDGTMQL5A&forceAllTransforms=true&modules=commonjs&shippedProposals=false&evaluate=false&fileSize=false&timeTravel=true&sourceType=script&lineWrap=true&presets=env&prettier=false&targets=&version=7.27.0&externalPlugins=&assumptions=%7B%7D)). I am good with the current status. 

To improve the time travel performance, maybe we can lazy highlight the immediate step output?

Reply:
Do you mean not highlighting when the user switches tabs, but highlighting after a few hundred milliseconds without new switching?

Reply:
I recorded some gifs to show the difference visually.
Note the difference between the highlighted themes in normal and fast mode.

main
![7f68d733-63ac-44df-9905-2ad31c53ed18](https://github.com/user-attachments/assets/6d038c14-3f43-4ee8-9732-db0ce14914a6)
normal
![d53cf2fa-92eb-441c-a0bf-6dbcedc65e64](https://github.com/user-attachments/assets/f2a34449-965a-4f98-becb-2f6bce96d2bd)
fastMode
![3660f24b-24cf-4b21-bab7-2536ddb6619a](https://github.com/user-attachments/assets/c2d74326-44c6-484f-a934-c5cd2d6119a5)

Reply:
Oh I mean the page load time: the time spent between opening the URL until the highlighted code is rendered.

The time spent switching tab does seem a bit slower, but I think it is still acceptable.

Aside: Now comparing this branch vs main, I do notice that the editor font size of monaco (12px) is smaller in than codemirror (0.875rem), which is also the font-size of other GUI text. Could you update the font-size of monaco?

Reply:
Unfortunately monaco only supports px, I tried modifying the css properties but that resulted in an exception.

Reply:
OK, in that case we should set it to 14px then.